src/main/flight/imu.c

```diff
@@ -457,41 +457,78 @@ static void imuCalculateEstimatedAttitude(float dT)
     bool useCOG = false;
 
 #if defined(USE_GPS)
-    if (STATE(FIXED_WING)) {
-        bool canUseCOG = sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat >= 6 && gpsSol.groundSpeed >= 300;
+    bool canUseCOG = false;
+    uint16_t groundCourse;
+    if (sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat >= 6) {
+        // TODO: Use GPS heading if the real compass fails during a flight
+        switch (imuConfig()->use_gps_heading) {
+            case GPS_HEADING_AUTO:
+                canUseCOG = STATE(FIXED_WING);
+                break;
+            case GPS_HEADING_ENABLED:
+                canUseCOG = true;
+                break;
+            case GPS_HEADING_DISABLED:
+                // Already false
+                break;
+        }
 
+        // If COG is now enabled, do platform dependant sanity checks and calculate it
         if (canUseCOG) {
-            if (gpsHeadingInitialized) {
-                // Use GPS heading if error is acceptable or if it's the only source of heading
-                if (ABS(gpsSol.groundCourse - attitude.values.yaw) < DEGREES_TO_DECIDEGREES(MAX_GPS_HEADING_ERROR_DEG) || !canUseMAG) {
-                    courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
-                    useCOG = true;
+            if (STATE(FIXED_WING)) {
+                // Fixed wing always flies forward, so we require a 3m/s speed and use
+                // the GPS groundCourse without any rotations
+                canUseCOG = gpsSol.groundSpeed >= 300;
+                groundCourse = gpsSol.groundCourse;
+            } else {
+                // MR might fly on any direction, but it tends to fly in the direction the
+                // head it's tilted to. To compensate for this, we rotate gpsSol.groundCourse
+                // by the tilt direction. Pitch and roll angles must fall in the [10, 90) so
+                // the data doesn't get polluted during flips or rolls.
+                // For now, we also require a harcoded speed of 3m/s, but we
+                // should adjust this depending on the maximum speed in modes which do limit
+                // it (e.g. RTH).
+                const uint16_t minAttitudeAngle = DECIDEGREES_TO_DEGREES(10);
```

Comment:
You can check `calculateCosTiltAngle()` which is essentially generalized tilt `cos(roll)*cos(pitch)`